File: cellularNetwork.py
def first_at_least(arr, value):
    lo = 0
    hi = len(arr)
    while lo < hi:
        mid = (lo + hi) // 2
        if arr[mid] >= value:
            hi = mid
        else:
            lo = mid + 1
    return lo

n, m = map(int, input().split())
cities = []
towers = []

for _ in range(n):
    city = int(input())
    cities.append(city)

for _ in range(m):
    tower = int(input())
    towers.append(tower)

minR = 0
for i in range(n):
    tower_right = first_at_least(towers, cities[i])
    tower_left = tower_right - 1

    minR_for_this_city = float('inf')
    if tower_right < m:
        assert towers[tower_right] >= cities[i]
        minR_for_this_city = min(minR_for_this_city, towers[tower_right] - cities[i])
    if tower_left >= 0:
        assert towers[tower_left] <= cities[i]
        minR_for_this_city = min(minR_for_this_city, cities[i] - towers[tower_left])

    minR = max(minR, minR_for_this_city)

print(minR)

File: cellularNetwork.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Main {
    // returns the first index in the list that is >= value, or list.size() if no such index exists
    static int firstAtLeast(List<Integer> list, int value) {
        int lo = 0, hi = list.size();
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (list.get(mid) >= value)
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        List<Integer> cities = new ArrayList<>();
        List<Integer> towers = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            int city = sc.nextInt();
            cities.add(city);
        }

        for (int i = 0; i < m; i++) {
            int tower = sc.nextInt();
            towers.add(tower);
        }

        int minR = 0;
        for (int i = 0; i < n; i++) {
            int towerRight = firstAtLeast(towers, cities.get(i));
            int towerLeft = towerRight - 1;

            int minRForThisCity = 2_000_000_000;
            if (towerRight < m) {
                assert towers.get(towerRight) >= cities.get(i);
                minRForThisCity = Math.min(minRForThisCity, towers.get(towerRight) - cities.get(i));
            }
            if (towerLeft >= 0) {
                assert towers.get(towerLeft) <= cities.get(i);
                minRForThisCity = Math.min(minRForThisCity, cities.get(i) - towers.get(towerLeft));
            }

            minR = Math.max(minR, minRForThisCity);
        }

        System.out.println(minR);
    }
}
